### PR TITLE
[FLINK-10968][kubernetes] Implement TaskManager Entrypoint for Kubernetes.

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change
This PR is part of FLINK-9953(Active Kubernetes integration). Implement TaskManager Entrypoint for Kubernetes. `KubernetesTaskExecutorRunner` is the executable entry point for running a TaskExecutor in a Kubernetes pod. It will parse the dynamic configuration option in args and use the new configuration to run a task manager.

## Brief change log

* add `KubernetesTaskExecutorRunner`

## Verifying this change

It is just a entrypoint class. All the functionality is in `TaskManagerRunner`. Will test it manually.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		}
+
+		final Configuration dynamicProperties = ConfigurationUtils.createConfiguration(
+			clusterConfiguration.getDynamicProperties());
+		final Configuration configuration = GlobalConfiguration.loadConfiguration(
+			clusterConfiguration.getConfigDir(), dynamicProperties);
+
+		FileSystem.initialize(configuration, PluginUtils.createPluginManagerFromRootFolder(configuration));
+
+		final String resourceID = System.getenv().get(KubernetesResourceManager.ENV_RESOURCE_ID);
```
Comment (zhijiangW) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java]: nit: It might be better to introduce `ResourceManagerOptions#ENV_FLINK_CONTAINER_ID`, then the `KubernetesTaskExecutorRunner` and previous `YarnTaskExecutorRunner` can both reference it.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		Preconditions.checkArgument(resourceID != null,
+			"Resource ID variable %s not set", KubernetesResourceManager.ENV_RESOURCE_ID);
+
+		TaskManagerRunner.runTaskManager(configuration, new ResourceID(resourceID));
```
Comment (zhijiangW) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java]: Compared with the behavior in  `YarnTaskExecutorRunner`,  it would catch the exception to log and exit directly. 

I am curious that are there any considerations for throwing exceptions to upper layer and what is the expected behavior for handling exceptions from the upper layer?

Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java]: `TaskExecutorRunner` is the entrypoint to start a task manager. So if some exception happens, we should exit. I will change the current behavior.

**Comment by wangyang0918** [comment]
@zhijiangW Thanks for your review. I agree with that `KubernetesTaskExecutorRunner` and `YarnTaskExecutorRunner` have some duplicated codes. I will add a `AbstractTaskExecutorRunner` to cover the common codes.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java (hunk 1 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java]: I do not suggest to use the same env, `ResourceID` in yarn and k8s have different means, `ContainerID` for yarn and `PodName` for k8s. In k8s, there is also containerId which means docker container, it will confuse the user and devs.

**Comment by wangyang0918** [comment]
@zhijiangW I have addressed all your comments through the last 3 commits. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 3]
```diff
@@ -286,29 +285,10 @@ public static void main(String[] args) throws Exception {
...
-			final Throwable strippedThrowable = ExceptionUtils.stripException(t, UndeclaredThrowableException.class);
-			LOG.error("TaskManager initialization failed.", strippedThrowable);
-			System.exit(STARTUP_FAILURE_RETURN_CODE);
-		}
+		runTaskManagerSecurely(args, ResourceID.generate());
 	}
 
-	@VisibleForTesting
-	static Configuration loadConfiguration(String[] args) throws FlinkParseException {
+	public static Configuration loadConfiguration(String[] args) throws FlinkParseException {
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: This change should be put together with another commit "[hotfix][yarn] Use public methods of TaskManagerRunner to simply YarnTaskExecutorRunner codes", because it is actually required by that commit, not the scope of this commit.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java (hunk 1 above)
Comment (zhijiangW) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/taskmanager/KubernetesTaskExecutorRunner.java]: Thanks for the explanation. That makes sense for me.

**Comment by zhijiangW** [comment]
Thanks for the fix @wangyang0918 !

It is almost good for me now.  I am not quite sure whether it is necessary to supplement a unit test for `TaskManagerRunner.runTaskManagerSecurely` to verify the passed args are correctly loaded into configuration.  Because the new `KubernetesTaskExecutorRunner` relies on this method which seems not be covered by tests before.

In addition, if we can decouple this PR with previous `KubernetesResourceManager`, then it can be merged separately to not be blocked by precious PRs.  The only reference by this PR is `KubernetesResourceManager.ENV_RESOURCE_ID`.

**Comment by zhijiangW** [comment]
Another nit suggestion for commit format. It is better to have only one sentence for commit title and can further give more explanations in commit message if want.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 3 above)
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Make sense. I will squash the two commit to one.

**Comment by wangyang0918** [comment]
@zhijiangW Thanks for your quick response. I have updated the PR and the PR could be merges separately.

1. Squash the first two commits and update the commit message
2. Add a test for loading dynamic properties `testLoadDynamicalProperties`
3. Rename `ENV_RESOURCE_ID ` to `ENV_FLINK_POD_NAME` and move to `Constants`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java [hunk 4]
```diff
@@ -331,6 +311,25 @@ public static void runTaskManager(Configuration configuration, ResourceID resour
...
 
+	public static void runTaskManagerSecurely(String[] args, ResourceID resourceID) {
+		try {
+			final Configuration configuration = loadConfiguration(args);
+
+			FileSystem.initialize(configuration, PluginUtils.createPluginManagerFromRootFolder(configuration));
+
+			SecurityUtils.install(new SecurityConfiguration(configuration));
+
+			SecurityUtils.getInstalledContext().runSecured((Callable<Void>) () -> {
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: nit: can remove `(Callable<Void>)` I guess

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java [hunk 5]
```diff
@@ -97,22 +91,7 @@ private static void run(String[] args) {
...
-			final CommandLineParser<ClusterConfiguration> commandLineParser =
-				new CommandLineParser<>(new ClusterConfigurationParserFactory());
-
-			ClusterConfiguration clusterConfiguration = null;
-			try {
-				clusterConfiguration = commandLineParser.parse(args);
-			} catch (FlinkParseException e) {
-				LOG.error("Could not parse command line arguments {}.", args, e);
-				commandLineParser.printHelp(YarnTaskExecutorRunner.class.getSimpleName());
-				System.exit(1);
```
Comment (zhijiangW) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java]: It seems we changed the behavior here. The `System.exit(1)` is replaced by throwing exception after reusing `TaskManagerRunner.loadConfiguration` below. 
Another option is catching `TaskManagerRunner.loadConfiguration` to exit as previous way.

Comment (zhijiangW) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java]: Or we can further reuse `TaskManagerRunner#runTaskManagerSecurely` instead?

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/Constants.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.kubernetes.utils;
+
+/**
+ * Constants for kubernetes.
+ */
+public class Constants {
```
Comment (zhijiangW) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/Constants.java]: I am not quite sure whether it is worth defining this class for only one constant variable, or we can define this variable directly in `KubernetesTaskExecutorRunner` class. If you prefer to a separate class, then it is better to rename the class as KubernetetsConstants to give a specific semantic.

**Comment by zhijiangW** [comment]
Thanks for the fix @wangyang0918 . I left several additional comments before final approval.

**Comment by zhijiangW** [comment]
In addition, remove the dot at the end of commit title.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java (hunk 4 above)
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java]: Right. I will remove the `(Callable<Void>)`.

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java (hunk 5 above)
Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java]: I think we should not `System.exit(1)` here. We should use `INIT_ERROR_EXIT_CODE` instead. 

We just can not use `TaskManagerRunner#runTaskManagerSecurely` for `YarnTaskExecutorRunner`. Because the configuration needs to be updated and then use the updated configuration to start task executor. It is the unique logics of Yarn.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/Constants.java (hunk 6 above)
Comment (wangyang0918) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/Constants.java]: I will make this PR depends on #FLINK-10935. The class `Constants` will be created in that PR. We just add a new variable `ENV_FLINK_POD_NAME`. Does it make sense?

Comment (zhijiangW) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/Constants.java]: Makes sense for me

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java (hunk 5 above)
Comment (zhijiangW) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnTaskExecutorRunner.java]: Sorry I missed the below catch clause for catching this exception and exiting finally.
If so it makes sense for me.
